alf/algorithms/particle_vi_algorithm.py

```diff
@@ -0,0 +1,322 @@
+# Copyright (c) 2020 Horizon Robotics and ALF Contributors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""A generic generator."""
+
+import gin
+import numpy as np
+import torch
+
+from alf.algorithms.algorithm import Algorithm
+from alf.algorithms.mi_estimator import MIEstimator
+from alf.data_structures import AlgStep, LossInfo, namedtuple
+import alf.nest as nest
+from alf.networks import Network, EncodingNetwork
+from alf.tensor_specs import TensorSpec
+from alf.utils import common, math_ops
+from alf.utils.averager import AdaptiveAverager
+
+
+@gin.configurable
+class ParVIAlgorithm(Algorithm):
+    """ParVIAlgorithm
+
+    ParVIAlgorithm maintains a set of particles that keep chasing some target
+    distribution. Two particle-based variational inference (par_vi) methods 
+    are implemented:
```

Comment:
Try to make the format right. You can view the doc locally following the instruction: https://alf.readthedocs.io/en/latest/notes/preview_doc_website.html

Reply:
Viewed and updated here and elsewhere.